### ISSUE TITLE
QOS Filter does not handle IllegalStateException and never releases passes

### ISSUE THREAD

**Issue by AdrianFarmadin** [eclipse/jetty.project#4323]
**Jetty version**:  jetty-9.4.22.v20191022

**Java version**:
```
openjdk version "1.8.0_222"
OpenJDK Runtime Environment (AdoptOpenJDK)(build 1.8.0_222-b10)
OpenJDK 64-Bit Server VM (AdoptOpenJDK)(build 25.222-b10, mixed mode)
```

**OS type/version**: Amazon Linux AMI

**Description**:

I see 160 exceptions like this within the same second:
```
2019-11-14 16:37:34 UTC WARNING /communication
java.lang.IllegalStateException: s=WOKEN rs=EXPIRE os=OPEN is=IDLE awp=false se=false i=false al=4
	at org.eclipse.jetty.server.HttpChannelState.dispatch(HttpChannelState.java:589)
	at org.eclipse.jetty.server.AsyncContextState.dispatch(AsyncContextState.java:94)
	at org.eclipse.jetty.servlets.QoSFilter.doFilter(QoSFilter.java:229)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1604)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:545)
```

Then in an instant the QOS Filter blocks all requests on the communication servlet with 160 threads blocking like this:
```
"jetty-4004" #4004 daemon prio=5 os_prio=0 tid=0x00007f775400f000 nid=0x1b6ba waiting on condition [0x00007f73a33f1000]
   java.lang.Thread.State: WAITING (parking)
	at sun.misc.Unsafe.park(Native Method)
	- parking to wait for  <0x00007f804f308e30> (a java.util.concurrent.Semaphore$FairSync)
	at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.parkAndCheckInterrupt(AbstractQueuedSynchronizer.java:836)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.doAcquireSharedInterruptibly(AbstractQueuedSynchronizer.java:997)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquireSharedInterruptibly(AbstractQueuedSynchronizer.java:1304)
	at java.util.concurrent.Semaphore.acquire(Semaphore.java:312)
	at org.eclipse.jetty.servlets.QoSFilter.doFilter(QoSFilter.java:177)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1604)
```

It seems the QOS Filter does not handle the IllegalStateException and never calls " _passes.release();" in the finally block (it also leaves the loop btw.) ==> the semaphore is stuck and never can be accessed anymore.

See "finally" block in "doPost" of the QOSFilter:
```
 finally
        {
            if (accepted)
            {
                for (int p = _queues.length - 1; p >= 0; --p)
                {
                    AsyncContext asyncContext = _queues[p].poll();
                    if (asyncContext != null)
                    {
                        ServletRequest candidate = asyncContext.getRequest();
                        Boolean suspended = (Boolean)candidate.getAttribute(_suspended);
                        if (Boolean.TRUE.equals(suspended))
                        {
                            candidate.setAttribute(_resumed, Boolean.TRUE);
                            asyncContext.dispatch();
                            break;
                        }
                    }
                }
                _passes.release();
            }
        }
```

**Comment by szymonbultrowicz** [comment]
Hey, any update here? It's currently blocking us from upgrading the Jetty version as it's a critical issue from our standpoint.

**Comment by janbartel** [comment]
@AdrianFarmadin or @szymonbultrowicz  do either of you have a test bed that can reproduce the problem? I'm going to make a PR with a fix, but it would be great if it could be tested.

**Comment by janbartel** [comment]
Merged to 9.4.x branch.
